File: blogger/backend/src/main/java/com/blog/blogger/service/ReportService.java
package com.blog.blogger.service;

import com.blog.blogger.dto.CreateReportDTO;
import com.blog.blogger.dto.ReportResponseDTO;
import com.blog.blogger.dto.UpdateReportDTO;
import com.blog.blogger.models.*;
import com.blog.blogger.repository.ReportRepository;
import com.blog.blogger.repository.PostRepository;
import com.blog.blogger.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class ReportService {
    
    @Autowired
    private ReportRepository reportRepository;
    
    @Autowired
    private PostRepository postRepository;
    
    @Autowired
    private UserRepository userRepository;
    
    /**
     * Create a simple report
     */
    @Transactional
    public ReportResponseDTO createReport(CreateReportDTO dto, User reporter) {
        // Find the post
        Post post = postRepository.findById(dto.getPostId())
                .orElseThrow(() -> new RuntimeException("Post not found"));
        
        // Optional: Prevent duplicate reports from same user
        // if (reportRepository.existsByReporterAndPost(reporter, post)) {
        //     throw new RuntimeException("You already reported this post");
        // }
        
        // Create report
        Report report = new Report(reporter, post, dto.getMessage());
        report = reportRepository.save(report);
        
        return new ReportResponseDTO(report);
    }
    
    /**
     * Get all reports (admin)
     */
    public List<ReportResponseDTO> getAllReports() {
        return reportRepository.findAll().stream()
                .map(ReportResponseDTO::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get unresolved reports (admin dashboard)
     */
    public List<ReportResponseDTO> getUnresolvedReports() {
        return reportRepository.findByResolvedFalse().stream()
                .map(ReportResponseDTO::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Get user's own reports
     */
    public List<ReportResponseDTO> getUserReports(User user) {
        return reportRepository.findByReporter(user).stream()
                .map(ReportResponseDTO::new)
                .collect(Collectors.toList());
    }
    
    /**
     * Update report (mark as resolved)
     */
    @Transactional
    public ReportResponseDTO updateReport(Long reportId, UpdateReportDTO dto) {
        Report report = reportRepository.findById(reportId)
                .orElseThrow(() -> new RuntimeException("Report not found"));
        
        report.setResolved(dto.isResolved());
        report.setAdminNotes(dto.getAdminNotes());
        
        report = reportRepository.save(report);
        return new ReportResponseDTO(report);
    }
    
    /**
     * Delete report
     */
    @Transactional
    public void deleteReport(Long id) {
        reportRepository.deleteById(id);
    }
    
    /**
     * Count unresolved (for admin badge)
     */
    public long countUnresolvedReports() {
        return reportRepository.countByResolvedFalse();
    }
}

File: blogger/frontend/src/app/core/services/report.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Report, CreateReportDTO, UpdateReportDTO } from '../models/report.model';

@Injectable({
  providedIn: 'root'
})
export class ReportService {
  private apiUrl = 'http://localhost:8080/auth/reports';

  constructor(private http: HttpClient) {}

  /**
   * User: Create a report for a post
   */
  createReport(reportData: CreateReportDTO): Observable<Report> {
    return this.http.post<Report>(this.apiUrl, reportData);
  }

  /**
   * User: Get my reports
   */
  getMyReports(): Observable<Report[]> {
    return this.http.get<Report[]>(`${this.apiUrl}/my`);
  }

  /**
   * Admin: Get all reports
   */
  getAllReports(): Observable<Report[]> {
    return this.http.get<Report[]>(`${this.apiUrl}/admin/all`);
  }

  /**
   * Admin: Get unresolved reports
   */
  getUnresolvedReports(): Observable<Report[]> {
    return this.http.get<Report[]>(`${this.apiUrl}/admin/unresolved`);
  }

  /**
   * Admin: Get count of unresolved reports
   */
  getUnresolvedCount(): Observable<{ count: number }> {
    return this.http.get<{ count: number }>(`${this.apiUrl}/admin/count-unresolved`);
  }

  /**
   * Admin: Update report (mark as resolved, add notes)
   */
  updateReport(id: number, updateData: UpdateReportDTO): Observable<Report> {
    return this.http.put<Report>(`${this.apiUrl}/admin/${id}`, updateData);
  }

  /**
   * Admin: Delete report
   */
  deleteReport(id: number): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${this.apiUrl}/admin/${id}`);
  }
}
